models/network.py

```diff
@@ -0,0 +1,185 @@
+from functools import partial
+import torch
+import torch.nn as nn
+
+
+fn_weight_init = {
+    'normal': nn.init.xavier_normal_,
+    'orthogonal': nn.init.orthogonal_,
+}
+
+
+fn_activation = {
+    'relu': nn.ReLU,
+    'tanh': nn.Tanh,
+}
+
+
+def _weight_init(m, init_method='normal'):
+    if isinstance(m, nn.Linear):
+        fn_weight_init[init_method](m.weight.data)
+
+        if m.bias is not None:
+            nn.init.constant_(m.bias.data, 1e-3)
+
+
+def _make_layers(dims, activation='relu', final_activation=False):
+    layers = []
+    for i in range(len(dims) - 1):
+        layers.append(nn.Linear(dims[i], dims[i + 1]))
+        layers.append(fn_activation[activation]())
+    if not final_activation:
+        layers = layers[:-1]
+    return nn.Sequential(*layers)
+
+
+class ProbabilisticModule(nn.Module):
+    """Base module with last linear layers for mean and logstd"""
+    def __init__(self, dim_in, dim_out, init_method='normal', fixed_logstd=None,
+                 logstd_min=-10, logstd_max=2):
+        super().__init__()
+        self.fixed_logstd = fixed_logstd
+        self.logstd_min = logstd_min
+        self.logstd_max = logstd_max
+
+        self.fc_mean = _make_layers([dim_in, dim_out])
+        self.fc_logstd = _make_layers([dim_in, dim_out])
```

Comment:
Q) Is it general to have separate layers for `mean` and `log_std`? One code that I saw before used shared layers and made two layers for the last output so that one is for mean and the other is for std.

Reply:
The `ProbabilisticModule` itself is actually not used alone (though it can be), and only provides the common structure to the child classes. Please check out the classes inheriting this, e.g., `Encoder` first uses shared layers, then apply those separate layers.